### PR TITLE
Clarification of CSR specs

### PR DESCRIPTION
**Pull request by MikeOpenHWGroup**
Hi @davideschiavone and @Silabs-ArjanB.  This PR supports the CSR spec review/update I requested at the last VTG status meeting (on August 18).   I struggled with how best to present what I am looking for - it is not clear that a GitHub pull-request is the best way, but at least its reviewable.

You may find it helpful to clone both openhwgroup/core-v-docs (master) and MikeOpenHWGroup/core-v-docs (csr_update) and use **tkdiff** to see the deltas on `user_manual/source/control_status_registers.rst`.

Signed-off-by: Mike Thompson <mike@openhwgroup.org>

### CODE REVIEW

File: cores/cv32e40p/user_manual/source/control_status_registers.rst [hunk 1]
```diff
@@ -9,135 +9,149 @@ that were needed for the PULP system. The reason for this is that we
...
+
+The **Privilege** column indicates the access mode of a CSR.  The first letter
+indicates the lowest privilege level required to access the CSR.  Attempts to
+access a CSR with a higher privilege level than the core is currently running
+in will throw an illegal instruction exception.  This is largely a moot point
+for the CV32E40P as it only supports machine and debug modes.  The remaining
+letters indicate the read and/or write behavior of the CSR when accessed by
+the indicated or higher privilge level:
+
+* **RW**: CSR is **read-write**.  That is, csr instructions may write any value
```
Comment (Silabs-ArjanB) [inline cores/cv32e40p/user_manual/source/control_status_registers.rst:31]: csr -> CSR

File: cores/cv32e40p/user_manual/source/control_status_registers.rst [hunk 2]
```diff
@@ -9,135 +9,149 @@ that were needed for the PULP system. The reason for this is that we
...
+  and that value will be returned on a subsequent read (unless a side-effect
+  causes the core to change the CSR value).
+
+* **RO**: CSR is **read-only**.  Writes by csr instrucions have no effect.
```
Comment (Silabs-ArjanB) [inline cores/cv32e40p/user_manual/source/control_status_registers.rst:35]: csr instrucions -> CSR instructions

Comment (Silabs-ArjanB) [inline cores/cv32e40p/user_manual/source/control_status_registers.rst:35]: Replace text by: Writes by CSR instructions raise an illegal instructions exception.

File: cores/cv32e40p/user_manual/source/control_status_registers.rst [hunk 3]
```diff
@@ -9,135 +9,149 @@ that were needed for the PULP system. The reason for this is that we
...
+| 0x7C2         | ``lpcount0``      | URW       | ``PULP_XPULP`` = 1  | Hardware Loop 0 Counter.                                |
++---------------+-------------------+-----------+---------------------+---------------------------------------------------------+
+| 0x7C4         | ``lpstart1``      | URW       | ``PULP_XPULP`` = 1  | Hardware Loop 1 Start.                                  |
++---------------+-------------------+-----------+---------------------+---------------------------------------------------------+
+| 0x7C5         | ``lpend1``        | URW       | ``PULP_XPULP`` = 1  | Hardware Loop 1 End.                                    |
++---------------+-------------------+-----------+---------------------+---------------------------------------------------------+
+| 0x7C6         | ``lpcount1``      | URW       | ``PULP_XPULP`` = 1  | Hardware Loop 1 Counter.                                |
++---------------+-------------------+-----------+---------------------+---------------------------------------------------------+
+| 0x006         | ``fprec``         | URW       | ``PULP_XPULP`` = 1  | Custom flag which controls the precision and latency    |
```
Comment (Silabs-ArjanB) [inline cores/cv32e40p/user_manual/source/control_status_registers.rst:68]: Condition is FPU = 1 && PULP_XPULP =1 (as correctly documented at the fprec description itself)

File: cores/cv32e40p/user_manual/source/control_status_registers.rst [hunk 4]
```diff
@@ -159,70 +173,87 @@ Table 7: Control and Status Register Map
...
 
+CSR Descriptions
+-----------------
+
+What follows is a detailed definition of each of the CSRs listed above.  The
+**Mode** column defines the access mode behavior of each bit field when
+accessed by the privilege level specified in Table 7 (or a higher privilege
+level):
+
+* **RO**: **read-only** fields are not affect by csr write instructions.  Such
```
Comment (Silabs-ArjanB) [inline cores/cv32e40p/user_manual/source/control_status_registers.rst:184]: csr -> CSR

File: cores/cv32e40p/user_manual/source/control_status_registers.rst [hunk 5]
```diff
@@ -261,26 +292,26 @@ Reset Value: 0x0000_0000
...
 CSR Address: 0x006 (only present if ``FPU`` = 1 and ``PULP_XPULP`` = 1)
 
 Reset Value: 0x0000_0000
 
 +-------------+-----------+----------------------------------------------------------------------------------+
-|   Bit #     |   R/W     | Description                                                                      |
+|   Bit #     |  Mode     | Description                                                                      |
 +=============+===========+==================================================================================+
-| 31:5        | R/W       | Writes are ignored; reads return 0.                                              |
+| 31:5        | RW        | Writes are ignored; reads return 0.                                              |
```
Comment (Silabs-ArjanB) [inline cores/cv32e40p/user_manual/source/control_status_registers.rst:312]: RO

File: cores/cv32e40p/user_manual/source/control_status_registers.rst [hunk 6]
```diff
@@ -362,30 +394,32 @@ Reset Value: 0x0000_1800
...
   +-------------+-----------+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
 
 Machine ISA (``misa``)
-----------------------
+~~~~~~~~~~~~~~~~~~~~~~
 
 CSR Address: 0x301
 
-Reset Value: 0x0000_0000
+Reset Value: 0x4080_1104
```
[the remark this answers is not in the document]
Comment (Silabs-ArjanB) [reply cores/cv32e40p/user_manual/source/control_status_registers.rst:416]: Actually it is a defined value. We need to merge https://github.com/openhwgroup/core-v-docs/pull/132 before this pull request.

File: cores/cv32e40p/user_manual/source/control_status_registers.rst [hunk 7]
```diff
@@ -755,62 +807,60 @@ Reset Value: 0x0000_0003
...
 
-+----------+
-| 31 : 0   |
-+==========+
-| DPC      |
-+----------+
++-------------+-----------+-------------------------------------------------------------------------------------------------+
+|   Bit #     |   Mode    |   Description                                                                                   |
++=============+===========+=================================================================================================+
+| 31:0        | RO        | DPC                                                                                             |
```
Comment (Silabs-ArjanB) [inline cores/cv32e40p/user_manual/source/control_status_registers.rst:841]: 31:1 is RW, bit 0 is RO

File: cores/cv32e40p/user_manual/source/control_status_registers.rst [hunk 8]
```diff
@@ -819,50 +869,47 @@ Reset Value: 0x0000_0000
 Detailed:
 
 +-------+------+------------------------------------------------------------------+
-| Bit#  | R/W  | Description                                                      |
+| Bit#  | Mode | Description                                                      |
 +=======+======+==================================================================+
-| 31:0  | R/W  | 0                                                                |
+| 31:0  | RO   | The lower 32 bits of the 64 bit machine mode cycle counter.      |
```
Comment (Silabs-ArjanB) [inline cores/cv32e40p/user_manual/source/control_status_registers.rst:874]: RW

File: cores/cv32e40p/user_manual/source/control_status_registers.rst [hunk 9]
```diff
@@ -819,50 +869,47 @@ Reset Value: 0x0000_0000
...
 
-+-------+------+------------------------------------------------------------------+
-| Bit#  | R/W  | Description                                                      |
-+=======+======+==================================================================+
-| 31:0  | R/W  | 0                                                                |
-+-------+------+------------------------------------------------------------------+
++-------+------+---------------------------------------------------------------------------+
+| Bit#  | Mode | Description                                                               |
++=======+======+===========================================================================+
+| 31:0  | RO   | The lower 32 bits of the 64 bit machine mode instruction retired counter. |
```
Comment (Silabs-ArjanB) [inline cores/cv32e40p/user_manual/source/control_status_registers.rst:890]: RW

File: cores/cv32e40p/user_manual/source/control_status_registers.rst [hunk 10]
```diff
@@ -871,50 +918,47 @@ Reset Value: 0x0000_0000
 Detailed:
 
 +-------+------+------------------------------------------------------------------+
-| Bit#  | R/W  | Description                                                      |
+| Bit#  | Mode | Description                                                      |
 +=======+======+==================================================================+
-| 31:0  | R/W  | 0                                                                |
+| 31:0  | RO   | The upper 32 bits of the 64 bit machine mode cycle counter.      |
```
Comment (Silabs-ArjanB) [inline cores/cv32e40p/user_manual/source/control_status_registers.rst:923]: RW

File: cores/cv32e40p/user_manual/source/control_status_registers.rst [hunk 11]
```diff
@@ -871,50 +918,47 @@ Reset Value: 0x0000_0000
...
 
-+-------+------+------------------------------------------------------------------+
-| Bit#  | R/W  | Description                                                      |
-+=======+======+==================================================================+
-| 31:0  | R/W  | 0                                                                |
-+-------+------+------------------------------------------------------------------+
++-------+------+---------------------------------------------------------------------------+
+| Bit#  | Mode | Description                                                               |
++=======+======+===========================================================================+
+| 31:0  | RO   | The upper 32 bits of the 64 bit machine mode instruction retired counter. |
```
Comment (Silabs-ArjanB) [inline cores/cv32e40p/user_manual/source/control_status_registers.rst:939]: RW

File: cores/cv32e40p/user_manual/source/control_status_registers.rst [hunk 12]
```diff
@@ -987,22 +1030,20 @@ Table 14: MHARTID
...
-  |   Bit #     |   R/W     |   Description                                                                                                 |
+  |   Bit #     |   Mode    |   Description                                                                                                 |
   +=============+===========+===============================================================================================================+
-  | 31 : 2      |   R/W     | BASE: The trap-handler base address, always aligned to 256 bytes, i.e., utvec[7:2] is always set to 0.        |
+  | 31 : 2      |   RW      | BASE: The trap-handler base address, always aligned to 256 bytes, i.e., utvec[7:2] is always set to 0.        |
   +-------------+-----------+---------------------------------------------------------------------------------------------------------------+
-  |  1 : 0      |   R/W     | MODE: 00 = direct mode, 01 = vectored mode. Writes to bit 1 are ignored and this bit always reads as 0.       |
+  |  1 : 0      |   RW      | MODE: 00 = direct mode, 01 = vectored mode. Writes to bit 1 are ignored and this bit always reads as 0.       |
```
Comment (Silabs-ArjanB) [inline cores/cv32e40p/user_manual/source/control_status_registers.rst:1037]: bit 1 is RO

**Review by Silabs-ArjanB — CHANGES_REQUESTED** [review]
- Although the RO, RW distinction is clear, the official RISC-V spec nomenclature (WPRI, WLRL, WARL) adds info that would otherwise be missing from our spec. I think we should use RO only if the entire CSR is marked as R. In our case we actually will have many WARL (0) bitfields (with the 0 indicating that the legal value read will always be 0). So, in short, I prefer we use WARL (0) in many cases where we state RO below. We would still use RW for fields that are truly read-write.
- Some fields have a wrongly marked mode.
- Please only merge this one after the other outstanding pull requests as this one will lead to many merge conflicts.

File: cores/cv32e40p/user_manual/source/control_status_registers.rst (hunk 1 above)
Comment (MikeOpenHWGroup) [reply cores/cv32e40p/user_manual/source/control_status_registers.rst:31]: I used lower case here deliberately as I was speaking of an instruction (e.g. csrrw), not a CSR.   That is not clear from what I wrote, so I made an update.

File: cores/cv32e40p/user_manual/source/control_status_registers.rst (hunk 2 above)
Comment (MikeOpenHWGroup) [reply cores/cv32e40p/user_manual/source/control_status_registers.rst:35]: Dito.

Comment (MikeOpenHWGroup) [reply cores/cv32e40p/user_manual/source/control_status_registers.rst:35]: You are right that this is what the priviliged-spec calls for (Section 2.1).  However, I do not see this in simulation.

Comment (Silabs-ArjanB) [reply cores/cv32e40p/user_manual/source/control_status_registers.rst:35]: Then please raise it as an RTL bug indeed

File: cores/cv32e40p/user_manual/source/control_status_registers.rst (hunk 3 above)
Comment (MikeOpenHWGroup) [reply cores/cv32e40p/user_manual/source/control_status_registers.rst:68]: Fixed.

File: cores/cv32e40p/user_manual/source/control_status_registers.rst (hunk 4 above)
Comment (MikeOpenHWGroup) [reply cores/cv32e40p/user_manual/source/control_status_registers.rst:184]: Fixed

File: cores/cv32e40p/user_manual/source/control_status_registers.rst (hunk 5 above)
Comment (MikeOpenHWGroup) [reply cores/cv32e40p/user_manual/source/control_status_registers.rst:312]: Fixed (good catch).

File: cores/cv32e40p/user_manual/source/control_status_registers.rst (hunk 7 above)
Comment (MikeOpenHWGroup) [reply cores/cv32e40p/user_manual/source/control_status_registers.rst:841]: Fixed

File: cores/cv32e40p/user_manual/source/control_status_registers.rst (hunk 8 above)
Comment (MikeOpenHWGroup) [reply cores/cv32e40p/user_manual/source/control_status_registers.rst:874]: Yes, you are right.  Fixed.

File: cores/cv32e40p/user_manual/source/control_status_registers.rst (hunk 9 above)
Comment (MikeOpenHWGroup) [reply cores/cv32e40p/user_manual/source/control_status_registers.rst:890]: Dito.

File: cores/cv32e40p/user_manual/source/control_status_registers.rst (hunk 12 above)
Comment (MikeOpenHWGroup) [reply cores/cv32e40p/user_manual/source/control_status_registers.rst:1037]: Yeah, I did not make any attempt to update the User Mode CSRs, but have made this correction.

**Comment by MikeOpenHWGroup** [comment]
Hi @Silabs-ArjanB, thanks for the detailed reivew.  I have pushed in an update with fixes to the issues you flagged.   A couple of points about your general comments:

> Although the RO, RW distinction is clear, the official RISC-V spec nomenclature (WPRI, WLRL, WARL) adds info that would
> otherwise be missing from our spec. I think we should use RO only if the entire CSR is marked as R. In our case we actually
> will have many WARL (0) bitfields (with the 0 indicating that the legal value read will always be 0). So, in short, I prefer we use
> WARL (0) in many cases where we state RO below. We would still use RW for fields that are truly read-write.

I am not sure I agree.  The field specifications from the ISA (e.g. WARL, etc.) describe how the software is required to interact with the hardware to ensure correct operation and avoid exceptions.  The user manual is a description of what the hardware does.  I submit that RO and RW designations are sufficient for that purpose.

> Let's first define the 'modes' that we will use, for example:
> 
>     * Read-only CSRs
>       o	RO bitfields (if we do not want to be specific about the read value)
>       o	RO (0) bitfields (if the bitfield is guaranteed 0 always)
> 
>     * Read/write CSRs
>       o	RW bitfields
>       o	WPRI bitfields
>       o	WLRL bitfields
>       o	WARL bitfields
>       o	WARL (0) bitfields

I do not believe this is necessary.  Defining RO (0) as a specific access mode seems redundant given that it is a RO CSR whose the power-on-reset value is 0x0.   For the RW CSRs I stand by my previous comment: the WPRI, etc. fields describe the software field specification, which is different from the hardware access mode behavior.  Having said that, its not an issue worth nit-picking over, so if you add this to the user manual I have no objection.

**Comment by MikeOpenHWGroup** [comment]
@Silabs-ArjanB, @davideschiavone, can we move this one forward (preferably to a merge)?

**Comment by Silabs-ArjanB** [comment]
Hi @MikeOpenHWGroup , I talked to Paul about the RO vs. WARL syntax and we are fine with sticking to RO as you suggested. So you can replace  the WARLs I already introduced for the misa bitfields with RO. You can also use the RO (0x0), RO (0x1), etc. syntax in case you know the actual read value (as I did when using WARL for misa).

**Comment by MikeOpenHWGroup** [comment]
Will do.

**Comment by Silabs-ArjanB** [comment]
Hi @MikeOpenHWGroup A number of documentation fixes are gated by this pull request (and its conflicting file). Could you please resolve the conflicts?

**Comment by MikeOpenHWGroup** [comment]
Hi @Silabs-ArjanB, sorry for the delay - I've been side-lined by DV plan reviews.   An update to the pull-request has been used in and I've resolved the conflicts with the master.

**Review by Silabs-ArjanB — CHANGES_REQUESTED** [review]
Hi @MikeOpenHWGroup Please change the misa reset value back to 'defined', then we are good to go.

**Comment by MikeOpenHWGroup** [comment]
Done.

**Pull request merged**